src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <deque>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "src/stirling/source_connectors/socket_tracer/protocols/common/interface.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/common/timestamp_stitcher.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h"
+
+
+namespace px {
+namespace stirling {
+namespace protocols {
+
+template <>
+inline RecordsWithErrorCount<mqtt::Record> StitchFrames(std::deque<mqtt::Message>* req_messages,
+                                                        std::deque<mqtt::Message>* resp_messages,
+                                                        NoState* /*state*/) {
+  return StitchMessagesWithTimestampOrder<mqtt::Record>(req_messages, resp_messages);
```

Comment:
The stitcher should be the next PR after the parser changes, so I don't want to dive into this too much now. However, it seems that message ordering is only guaranteed within a given QoS level. I think using `StitchMessagesWithTimestampOrder` will result in invalid protocol traces if a client used multiple QoS levels.

<img width="1483" alt="Screenshot 2023-11-01 at 8 54 26 AM" src="https://github.com/pixie-io/pixie/assets/5855593/40a5dcbe-0092-4b4c-8403-bc469e756cdd">

In the next PR we will add tests for the Stitcher and we can model that situation. We will likely need to perform a similar process as `StitchMessagesWithTimestampOrder`, making sure that we only match frames within the same QoS since that will guarantee its [assumptions hold](https://github.com/pixie-io/pixie/blob/b6cf0ac3e4723a409b41cc0b603f3b8ddc1001b3/src/stirling/source_connectors/socket_tracer/protocols/common/timestamp_stitcher.h#L32-L35). This is why it doesn't work with HTTP pipelining.

We can either use the QoS field as the map key to our new stitcher interface. This hasn't been documented yet as it's in the process of being merged (#1716) or we can leverage the protocol state to make sure we have the ordering correct. My initial thinking is that the former would be best, but we will have to see.

Reply:
Makes sense. I had initially used `StitchMessagesWithTimestampOrder` as a placeholder as I was not sure what would work. I will focus more on the stitcher in the stitcher PR. Thanks for the additional context, it's very helpful.